### DISCUSSION TITLE
Give me direction to choose between TurboModule and Fabric Native Component

### DISCUSSION THREAD

**Discussion by plibr** [reactwg/react-native-new-architecture#185] [Q&A] [answered]
would you give me direction to research, I need epub reader page in my RN app.

Q1. What should I use for it, TurboModule or Fabric Native Component?

Q2. It seems that Swift is not able to work directly in New Architecture. How to create communication with packages that written in Swift like, EPUBKit, FolioReaderKit?

Q3. Is it suitable to install EPUBKit or FolioReaderKit in for example RTNCalculator/ios directory?

**Comment by cortinico** [comment, marked as answer]
> Q1. What should I use for it, TurboModule or Fabric Native Component?

If it has to show anything on screen: then it's a Native Component.
If it has **not**: then it's a TurboModule.

You might also need to use both depending on what your application is doing.

> Q2. It seems that Swift is not able to work directly in New Architecture. How to create communication with packages that written in Swift like, EPUBKit, FolioReaderKit?

This has already been discussed here:
https://github.com/reactwg/react-native-new-architecture/discussions/143
https://github.com/reactwg/react-native-new-architecture/discussions/40
https://github.com/reactwg/react-native-new-architecture/discussions/57

> Q3. Is it suitable to install EPUBKit or FolioReaderKit in for example RTNCalculator/ios directory?

@cipolleschi can help you further here

**Comment by cipolleschi** [comment]
> Q3. Is it suitable to install EPUBKit or FolioReaderKit in for example RTNCalculator/ios directory?

TL;DR: yes, of course. You can install the dependency and create a wrapper around it.

The way to go would be:
1. Create your library following the proper guide FabricComponent/TurboModule
2. When defining the podspec, add a dependency on EPUBKit or FolioReaderKit
3. Write your component/module interface to call the APIs of the wrapper.

An example of something silimar is [this library](https://github.com/TrueLayer/truelayer-react-native-sdk). This is a React Native wrapper around a payment library which is implemented natively in Kotlin/Swift.
